compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java

```diff
@@ -187,13 +196,26 @@ void rewriteServiceVariable(BLangService service, SymbolEnv env, BLangBlockStmt
         }
     }
 
+    private BType getListenerType(BType type) {
+        if (type.tag == TypeTags.UNION) {
+            for (BType memberType : ((BUnionType) type).getMemberTypes()) {
+                if (types.checkListenerCompatibility(memberType)) {
+                    return memberType;
+                }
+            }
+        }
+        return type;
+    }
+
     private void addMethodInvocation(Location pos, BLangSimpleVarRef varRef, BInvokableSymbol methodRefSymbol,
                                      List<BLangExpression> args,
                                      BlockNode body) {
         // Create method invocation
         final BLangInvocation methodInvocation =
                 ASTBuilderUtil.createInvocationExprForMethod(pos, methodRefSymbol, args, symResolver);
-        methodInvocation.expr = varRef;
+        BLangCheckedExpr listenerCheckExpr = ASTBuilderUtil.createCheckExpr(pos, varRef, getListenerType(varRef.type));
+        listenerCheckExpr.equivalentErrorTypeList.add(symTable.errorType);
```

Comment:
No impact right now, but should we add the actual error types here? We can do this later maybe.